### PR TITLE
Fix invalid padding length in reflection_pad2d_backward

### LINKED ISSUE #194048: [CPU] reflection_pad2d_backward reads past short padding lists
### 🐛 Describe the bug

The schema expects four padding values, but direct boxed dispatch accepts a
short list. The CPU implementation then reads `padding[0]` through
`padding[3]` without first checking the list length.

### Reproduction

```python
import torch

x = torch.randn([1, 1, 1], dtype=torch.float32)
grad = torch.randn([1, 1, 1], dtype=torch.float32)
torch.ops.aten.reflection_pad2d_backward(grad, x, [0])
```

### Error message

The current ASAN CPU build reports an 8-byte heap-buffer-overflow read directly
after the one-element padding vector. The following is a verbatim excerpt;
only absolute build-root path prefixes were redacted.

```text
==9==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x602000e1ce98 at pc 0x7fde792f6e88 bp 0x7ffc3526f650 sp 0x7ffc3526f648
READ of size 8 at 0x602000e1ce98 thread T0
==9==WARNING: Can't read from symbolizer at fd 3
==9==WARNING: Can't write to symbolizer at fd 6
    #0 0x7fde792f6e87  (<python-env>/lib/python3.11/site-packages/torch/lib/libtorch_cpu.so+0x4838e87)
    #1 0x7fde792f7768 in at::native::reflection_pad2d_backward_cpu(at::Tensor const&, at::Tensor const&, c10::ArrayRef<long>) <pytorch-source>/aten/src/ATen/native/ReflectionPad.cpp:366:3
    #2 0x7fde7e6054e2 in at::(anonymous namespace)::(anonymous namespace)::wrapper_CPU__reflection_pad2d_backward(at::Tensor const&, at::Tensor const&, c10::ArrayRef<c10::SymInt>) <pytorch-source>/build/aten/src/ATen/RegisterCPU_2.cpp:7822:10
    #3 0x7fde7e80f128 in c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, at::Tensor const&, c10::ArrayRef<c10::SymInt>), &at::(anonymous namespace)::(anonymous namespace)::wrapper_CPU__reflection_pad2d_backward(at::Tensor const&, at::Tensor const&, c10::ArrayRef<c10::SymInt>)>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, at::Tensor const&, c10::ArrayRef<c10::SymInt>>>::operator()(at::Tensor const&, at::Tensor const&, c10::ArrayRef<c10::SymInt>) <pytorch-source>/aten/src/ATen/core/boxing/impl/WrapFunctionIntoFunctor.h:17:12
0x602000e1ce98 is located 0 bytes after 8-byte region [0x602000e1ce90,0x602000e1ce98)
allocated by thread T0 here:
    #0 0x7fdeac3aa98d in operator new(unsigned long) <llvm-source>/compiler-rt/lib/asan/asan_new_delete.cpp:95:3
    #1 0x7fdea3d15939 in std::__new_allocator<c10::SymInt>::allocate(unsigned long, void const*) <toolchain>/include/c++/12/bits/new_allocator.h:137:27
    #2 0x7fdea3d158c0 in std::allocator_traits<std::allocator<c10::SymInt>>::allocate(std::allocator<c10::SymInt>&, unsigned long) <toolchain>/include/c++/12/bits/alloc_traits.h:464:20
    #3 0x7fdea3d1587f in std::_Vector_base<c10::SymInt, std::allocator<c10::SymInt>>::_M_allocate(unsigned long) <toolchain>/include/c++/12/bits/stl_vector.h:378:20
SUMMARY: AddressSanitizer: heap-buffer-overflow <pytorch-source>/aten/src/ATen/native/ReflectionPad.cpp:282:19 in at::native::(anonymous namespace)::reflection_pad2d_backward_out_template(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long>)
```

### Expected behavior

The operator should reject padding lists whose length is not four.

### Additional context

The schema declares `SymInt[4]`, but direct `torch.ops` dispatch accepts the
one-element Python list. `aten/src/ATen/native/ReflectionPad.cpp` then reads
`padding[0]` through `padding[3]` without first checking the list length.

### Versions

```text
PyTorch version: 2.11.0a0+git70d99e9
Git revision: 70d99e998b4955e0049d13a98d77ae1b14db1f45
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Linux (x86_64)
Libc version: glibc-2.35
Python version: 3.11.15 (64-bit runtime)
Python platform: Linux-7.0.0-29-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
Is XPU available: False
Is XNNPACK available: False

CPU architecture: x86_64
CPU model: Intel(R) Xeon(R) Platinum 8368 CPU @ 2.40GHz

Build instrumentation: AddressSanitizer
[pip3] numpy==2.4.6
[pip3] optree==0.19.1
[pip3] torch==2.11.0a0+git70d99e9
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @malfet

### PR DESCRIPTION
**Pull request by NishchalTendulkar**
Fixes #194048

### Summary

`reflection_pad2d_backward` assumes that the `padding` argument contains exactly four values. When a shorter padding list is passed, the implementation accesses padding elements before validating the size of the list, which can result in an out-of-bounds read and an invalid runtime error.

This PR adds an explicit validation to `reflection_pad2d_backward_out_template` to ensure that `padding` contains exactly four values before accessing its elements.

### Changes

- Added a `TORCH_CHECK` requiring `padding.size() == 4`.
- Added regression coverage for invalid padding lengths:
  - `()`
  - `(0,)`
  - `(0, 0)`
  - `(0, 0, 0)`
  - `(0, 0, 0, 0, 0)`
- Valid four-element padding remains unchanged.

### Testing

Ran the targeted regression test:

```text
python test/test_nn.py -v -k test_ReflectionPad_fails

test_ReflectionPad_fails_cpu (__main__.TestNNDeviceTypeCPU) ... ok

Ran 1 test in 0.009s
OK

### CODE REVIEW

File: aten/src/ATen/native/ReflectionPad.cpp [hunk 1]
```diff
@@ -272,6 +272,11 @@ void reflection_pad2d_out_template(
 void reflection_pad2d_backward_out_template(
     Tensor &grad_input, const Tensor &grad_output,
     const Tensor &input, IntArrayRef padding) {
+  TORCH_CHECK(
```
Comment (Skylion007) [inline aten/src/ATen/native/ReflectionPad.cpp]:
```suggestion
  TORCH_CHECK_VALUE(

File: test/test_nn.py [hunk 2]
```diff
@@ -8856,6 +8856,18 @@ def test_ReflectionPad_fails(self, device):
             inp = torch.randn(3, 3, 10, 10, 10, device=device)
             torch.ops.aten.reflection_pad2d(inp, (2, 2, 2, 2))
 
+        grad_output = torch.randn(1, 1, 1, device=device)
+        inp = torch.randn(1, 1, 1, device=device)
+
+        for padding in [(), (0,), (0, 0), (0, 0, 0), (0, 0, 0, 0, 0)]:
+            with self.assertRaisesRegex(
+                RuntimeError,
```
Comment (Skylion007) [inline test/test_nn.py]:
```suggestion
                ValueError,

**Comment by janeyx99** [comment]
how much does this impact perf? i agree it is nice to have validation but at some point there is a tradeoff with perf

**Comment by NishchalTendulkar** [comment]
Done. I compared the benchmark on the base revision and the updated branch. The median runtime was ~15.32 µs on the base and ~14.9–15.4 µs on the updated branch, so there doesn't appear to be any measurable performance regression.
